### ISSUE TITLE
Regex with a " in character class

### ISSUE THREAD

**Issue by mortenpi** [JuliaLang/julia#28261]
`match(r"[\\\"]", "\\")` does not match on 0.7 while `match(r"[\"\\]", "\\")` is fine. Unless I am messing up with escape sequences here somehow, [either should match as far as I can tell](https://regex101.com/r/dxPgCn/1).

Tried on 0.7.0-beta2.83 and 0.7.0-beta2.0. It works fine on 0.6, so looks like a regression.

**Comment by vtjnash** [comment]
not a bug, but the intended consequence of #22926

In regex, these two are equivalent (and non-matching), since regexes `[\"]` and `["]` are identical:
```julia
julia> match(r"[\"]", "\\")

julia> match(r"[\\\"]", "\\")
```

On v0.7, what you want is the regex `[\\"]` (or, equivalently, `[\\"]`):
```julia
julia> match(r"[\\\\\"]", "\\")
RegexMatch("\\")
```

But it looks like we're not printing the `regex` string correctly (it doesn't round-trip properly), which probably makes this even more confusing. I've got some of a fix for that locally and I'll continue working on cleaning that up and getting it into a PR.

**Comment by mortenpi** [comment]
> On v0.7, what you want is the regex `[\\"]`

That's what I assumed the first one to be, since I got the impression that backslashes don't need escaping in `@r_str` (my initial regex was more like `r"\\([\\\"])]"`, and the initial `\\` got me confused). The `@r_str` macro docstring is also outdated then if I understand correctly?

> Construct a regex, such as `r"^[a-z]*$"`, without interpolation and unescaping (except for quotation mark " which still has to be escaped).

It implies to me that the only escape sequence is `\"`.

**Comment by StefanKarpinski** [comment]
I think we need to rework a lot of the escaping behavior in custom string literals after that change. This means `r"..."` and `s"..."` in particular. This is fairly annoying and tedious to get right, unfortunately.

**Comment by wdebeaum** [comment]
#28175 seems related.

**Comment by StefanKarpinski** [comment]
I think we need to mark this as unstable in 1.0 because it needs to be fixed; that can be justified that it's a bugfix but it will likely break some code when we fix it.

**Comment by JeffBezanson** [comment]
I'm not sure how to fix this --- the escaping rules happen in the parser before we get to the custom string literal code, since otherwise we don't know when the string ends.

**Comment by StefanKarpinski** [comment]
Yeah, I'll have to look at it for a while and think; it may not be fixable.

**Comment by vtjnash** [comment]
I think it's just a printing issue, since it doesn't round-trip correctly. We can't to anything about the escaping rules, unless we want there to be a subset of regexes which cannot be expressed using the string macro (as was true in v0.6 and prior).

**Comment by StefanKarpinski** [comment]
If it's just a printing issue, that's even better since fixing them won't break people's code.

**Comment by mortenpi** [comment]
As far as I can tell the issue here is that the escaping logic is a bit counter-intuitive? But it is working as designed (except for the printing?). And, as was discussed at length in #22926, there might not really be any better solution.

> unless we want there to be a subset of regexes which cannot be expressed using the string macro (as was true in v0.6 and prior).

FWIW, one of the proposals in #22926 only excluded trailing backslashes, which would not be a problem  for regular expressions (and latex), since they are not allowed anyway.

**Comment by wdebeaum** [comment]
I'm sorry, what's an example of a regex that couldn't be expressed using the string macro in 0.6? I understand some strings couldn't be expressed with the raw string macro, but strings and regexes aren't 1:1. PCRE treats `\"` the same as `"`, so e.g. `Regex("a\\\"b")` matches the same as `Regex("a\"b")`.

**Comment by vtjnash** [comment]
The sequence `\"` (`Regex("\\\"")`) could not previously be given. `r"\""` would give `"` (`Regex("\"")`), while `r"\\\""` would give `\\"` (`Regex("\\\\\"")`). You're right that there's probably enough redundancy in regex that this may be immaterial (since 2n and 2n+1 `\` before " should give the same matches). If we want, we could precisely restore the old behavior by doubling the number of `\` that immediately precede any `"` character or the end of the string.
